### DISCUSSION TITLE
How do you do all this? :p

### DISCUSSION THREAD

**Discussion by apiel** [bangcorrupt/hacktribe#196] [General]
Hi,

Would be great to have more information on how you proceed to hack the Electribe, so maybe we could help you for adding more feature.
For example, how do you do to create new OSC. I was reading in some discussions that you use the original OSC and change some parameters, often also giving reference to the KingKORG. Are you first modifying the OSC on the KingKORG to have a preview of how it would sound like and then apply those change in the firmware? Can you maybe describe how the OSC are organize in the firmware? Are they located in a fix place? Do we need to edit them in binary (uint8, uint16...?)? Or would it even be possible to use Midi Sysex for that? How do set what the edit button change?
What would be amazing, once all this process is known, maybe someone could build a UI to create custom OSC, a bit like novation circuit... (the only problem I could see is how to have a live preview of what we changes...)

How about the IFX? Do you get the IFX from KingKORG?

From what you did so far, how much can the electribe be hacked? We can change OSC, IFX, groove, scale, adding sampling function to ES2. Do you see more possibilities compare to what you already did?

Thx, for your code, it already helped me a lot to make this (buggy stuff) https://github.com/apiel/electribe-ui

**Comment by bangcorrupt** [comment]
Hi, I don't have a KingKorg but the parameter guide and firmware binary have been useful.  KingKorg, electribe 2 and mini KP2s all use parts of the same sound engine in different ways.  The oscillators I added are just presets for the existing sound engine.  At some point I'll add python scripts to import/export oscillator presets via sysex, but there is not much more that can be changed without hacking the DSP firmware.  I can see how to have live editing of the underlying FX engine, so we could make new FX presets on the device in real-time, I just need to work out how to print to the screen.  

In the future I hope to share some source code that links into the patch, it's a lot of work to do it in a way that doesn't include partial decompilation of the factory firmware.  I need to re-work the whole hack from the start and fix some bugs before going much further, most of it is just hex edited straight into the binary.  I think the possibilities are endless, but at some point it makes more sense to work on a fully open source firmware from scratch.  It may be a while before there are any major changes, just got busy with real life.

**Reply by BKLronin** [reply]
+1 for OSS from scratch

**Reply by apiel** [reply]
But does it mean that you are modifying the original firmware using assembly (or something like this) through rizin to add more feature?
"import/export oscillator presets via sysex" is this already possible or do you have to modify the firmware as well?
I wish I could help, but I am already so busy with my own project(s) and I was on the point to sell my electribe 2, but now that I see your project, I am not so sure anymore :p

Open source firmware from scratch, would be a huge amount of work, especially to implement the synth engine... But I have to say, it would be like a dream to have such a project. At least a starting point, then people could take over to bring more features. What would be great is to build a kind of small framework similar to teensy audio library, that could easily run on the electribe, and then it would be up to the developers to design their own electribe groovebox. Hehe, it's good to dream :-)

**Comment by bangcorrupt** [comment]
At the start I was just hex editing the binary.  I've started re-working everything into an organised set of assembly source files, so at some point we can have functions for things like MIDI input or button presses that are completely decoupled from the factory firmware.

I'd like to make a custom firmware for the CPU that has an interface to the DSP sound engine.  Then we can carve the DSP firmware from the update file and keep the original sound engine.  There is also [ArduinoCore-blackfin](https://github.com/deanm1278/ArduinoCore-blackfin).  

Importing oscillator presets is not in the firmware yet.  All of the sysex hacks need reworking before  I add any more.
